python/dask_cudf/dask_cudf/sorting.py

```diff
@@ -0,0 +1,381 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+import math
+from collections.abc import Iterator
+from operator import getitem
+
+import cupy
+import numpy as np
+import tlz as toolz
+
+from dask.base import tokenize
+from dask.dataframe.core import DataFrame, Index, Series, _concat
+from dask.dataframe.shuffle import rearrange_by_column
+from dask.dataframe.utils import group_split_dispatch, hash_object_dispatch
+from dask.highlevelgraph import HighLevelGraph
+from dask.utils import M, digit, insert
+
+import cudf as gd
+from cudf.utils.dtypes import is_categorical_dtype
+
+
+def set_partitions_hash(df, columns, npartitions):
+    c = hash_object_dispatch(df[columns], index=False)
+    return np.mod(c, npartitions)
+
+
+def _shuffle_group(df, columns, stage, k, npartitions, ignore_index):
+    c = hash_object_dispatch(df[columns], index=False)
+    typ = np.min_scalar_type(npartitions * 2)
+    c = np.mod(c, npartitions).astype(typ, copy=False)
+    np.floor_divide(c, k ** stage, out=c)
+    np.mod(c, k, out=c)
+    return group_split_dispatch(
+        df, c.astype(np.int32), k, ignore_index=ignore_index
+    )
```

Comment:
:/ I'm 99% with you on this...

The current "hold up" is that the shuffling routine depends on the modulus logic performed between the initial hash calculation, and the splitting of the partition. During calls to `_shuffle_group` in `stage=0`, `partition_by_hash` can be used without any problems, for `stage>0`, you cannot get the correct offsets using a single call to `partition_by_hash`.

Asside from the need to modify the overall shuffling logic for multiple stages, the current implementation of `partition_by_hash` is also slower and more memory-hungry in cases that the index does not need to be preserved after the shuffle (like for set_index).  While `scatter_by_map` allows you to pass `keep_index=False`,  `partition_by_hash` will always include the index column.

Reply:
> the current implementation of `partition_by_hash` is also slower and more memory-hungry in cases that the index does not need to be preserved after the shuffle (like for set_index). While `scatter_by_map` allows you to pass `keep_index=False`, `partition_by_hash` will always include the index column.

We can easily fix this very quickly.


> The current "hold up" is that the shuffling routine depends on the modulus logic performed between the initial hash calculation, and the splitting of the partition. During calls to `_shuffle_group` in `stage=0`, `partition_by_hash` can be used without any problems. For `stage>0`, you cannot get the correct offsets using a single call to `partition_by_hash`.

So I understand up until the floor divide and second modulo operation come in. I'm unclear what `k` and `stage` are in this context and how the `floor_divide` followed by a second modulo gives different desirable behavior then just doing a hash partition with a smaller number of partitions.

Reply:
> I'm unclear what k and stage are in this context and how the floor_divide followed by a second modulo gives different desirable behavior then just doing a hash partition with a smaller number of partitions.

`k` is the number of splits you want to create as the output of this particular call to `_shuffle_group`, and that value depends on `npartitions` (the number of partitions desired after the multi-stage shuffle is finished) and the total number of stages.  My understanding is that: (1) The first modulus produces a column with the final partition for each row, (2) The floor-division aggregates the number of outputs to match the actual number of tasks that will be taking data for that particular stage of shuffling, (3) The final modulus makes sure there are only `k` splits generated.

I may be missunderstanding the logic a bit (I took a lot of time to work things out a few weeks back, and already forget the exact algorithm), but I believe the problem is actually the floor-divide.  It effectively means that you must use `partition_by_hash` to get `npartition` divisions, and then manually aggregate them.

Reply:
@kkraus14 - I will raise an issue regarding the `keep_index`/`ignore_index` option to `partition_by_hash`.  With that optimization in place, we *should* be able to use it for `_shuffle_group` (we can just aggregate splits together, after-the-fact, according to the same floor-division logic).

Reply:
From the looks of the implementation: we generate `n` **populated** output tables where `n` is the number of unique values in `c` and then add empty Tables to ensure we have `k` tables. From what I can tell we're not depending on any form of ordering or specific logic in the partitioning itself. Also from what I can tell, the only times `n` doesn't equal `k` is if you have very low cardinality data where the cardinality is below `k`.

I think this leaves us two avenues to explore:
- Use `partition_by_hash` with `k` which could potentially give us empty Tables in the middle of the list of output Tables (not sure if this is an issue or not).
- Calculate the number of unique values in `c` as `n` and use `partition_by_hash` with `n`, and then append empty Tables to the end of the list as needed.
    - It's not immediately clear to me if there's a more efficient way to calculate `n` then what we're currently doing with the `floor_divide` and `mod`.

Reply:
> From the looks of the implementation: we generate n populated output tables where...

Are you describing the `_shuffle_group` implementation, or `partition_by_hash`? If `_shuffle_group`, note that just using `partition_by_hash` with `k` will produce output equivalent to the following (correct?):

```python
def _shuffle_group(df, columns, stage, k, npartitions, ignore_index):
    c = hash_object_dispatch(df[columns], index=False)
    np.mod(c, k, out=c)
    return group_split_dispatch(df, c, k, ignore_index=ignore_index)
```

Unfortunately, this does not produce correct results after multi-stage shuffling (I will add a new test that exposes/catches this).

Reply:
I was trying to explain the `_shuffle_group` implementation but I may have gotten it incorrect :)

Reply:
I guess what I'm proposing is making some new function that looks like:
```
def group_split_dispatch(df, n, k, ignore_index):
    """
    n --> number of unique values in c
    """
    out_tables = partition_by_hash(df. n, ignore_index=ignore_index)
    if n < k:
        diff = k - n
        out_tables + ([Table()] * diff)
    return out_tables
```

Reply:
Right - I think I see what you are getting at.  I think you are focusing on some of the finer details of the implementation of `_shuffle_group`/`group_split_dispatch`, and I am still unsure how it will fit into the larger shuffle algorithm.

I may be overthinking/misunderstanding the higher-level algorithm, but I think the real problem is that your earlier hypothesis that "we're not depending on any form of ordering or specific logic in the partitioning itself" isn't quite true.  The dependency graph constructed for the shuffle makes specific assumptions about what data ends up in each output split of `_shuffle_group`.

The algorithm assumes that we know, a priori, the final partition for each row of the dask_cudf dataframe (It was originally designed for dask's `rearrange_by_column` function).  For each row, the final partition is just `<row-hash> % npartitions`.  When we are using a single-stage shuffle, the sub-graph (tree) used to construct partition `i` will simply take/concatenate the `i`th output of `_shuffle_group` from each of the original partitions.  When there are multiple stages, a similar tree is executed for each stage, but each time it will only take data from a specific a sub-set of partitions from the previous stage (`k` of them, I believe).  For this reason, it is important that the data is correctly associated with a specific partition for each stage - Otherwise, data can be literally "lost in the shuffle".

Reply:
> may be overthinking/misunderstanding the higher-level algorithm, but I think the real problem is that your earlier hypothesis that "we're not depending on any form of ordering or specific logic in the partitioning itself" isn't quite true. The dependency graph constructed for the shuffle makes specific assumptions about what data ends up in each output split of `_shuffle_group`.

So that's what would be captured by:
```
c = np.mod(c, npartitions).astype(typ, copy=False)
```
But then we immediately overwrite `c` in the following two lines which effectively removes the ability to map to the final output partition:
```
np.floor_divide(c, k ** stage, out=c)
np.mod(c, k, out=c)
```

If `npartitions` is say 4, but `k` is 2, your final map passed to the `scatter_to_tables` will only have at most `k` unique values. Now the data will absolutely differ: in the current implementation after the initial hash + modulo, values with 0 and 1 would end up in one output, values with 2 and 3 would end up in the other output, whereas if we just passed `k` to `partition_by_hash` then we get a different split of data. This shouldn't matter though as this is hash partitioning and guarantees that all similarly valued keys end up in the same output table number for each worker.

Reply:
FYI: we had an offline discussion surrounding this and came to a few conclusions.

1. This function is used for more than just hash partitioning, i.e. it's used in `set_index` where the output order of the partitions matters, so we can't just use a straight up `partition_by_hash` internal approach.
2. The slow part from the libcudf side in `scatter_to_tables` is the fact that we materialized new allocations for each output Table which then needed `N` passes over the data to populate. Instead we can return one contiguous Table and a vector of offsets that split up the Table, and we can do that much more efficiently with the Python side looking exactly the same.